content/docs/learn/typed-errors/from-either-to-raise.md

```diff
@@ -0,0 +1,280 @@
+      t.summarize()
+    }
+  }
+```
+
+Within the `Raise` DSL, the corresponding function is called `withError`.
+Following the discussion above, the `withError` function creates a new
+scope in which the error type is different. The first argument to
+`withError` describes how to transform failure values from the inner
+scope into the error type of the outer scope.
+
+```kotlin
+fun bar(n: Int): Either<Error, String> = either {
+  val s = f(n).bind()
+  val t = withError({ boo -> boo.toError() }) { h(s).bind() }
+  t.summarize()
+}
+```
+
+## More than one failure
+
+Sometimes you have not one, but many values for which you
+want to run a potentially-failing computation. Almost every library
+provides an "effectful map" (usually called `traverse`) to cover those cases.
+
+```kotlin
+fun foos(xs: List<Int>) = xs.traverse { foo(it) }
+```
+
+Another advantage of the `Raise` DSL is that you don't need all of those
+new combinators. The regular `map` on lists is enough — although you
+need to remember to `bind` the values to "inject" the errors.
+
+```kotlin
+fun foos(xs: List<Int>) = either {
+  xs.map { foo(it).bind() }
+  // alternatively
+  xs.map { foo(it) }.bindAll()
+}
+```
+
+This advantage is bigger than just dropping a few function names. Since
+you don't need specific effectful combinators, you are free to use any
+function operating on collections of elements, without restriction.
+
+:::tip Error accumulation
+
+By default `Raise` operates on a fail-first basis: once a failure is `raise`d,
+the whole block comes to a halt. If you are operating on a collection, this
+means that only the first error is reported. If you need to _accumulate_
+all errors that arise in some block, you need to switch to `mapOrAccumulate`.
+More information can be found in the 
+[Working with typed errors](working-with-typed-errors.md#accumulating-errors) page.
+
+:::
+
+## `Either` and `bind` no more
+
+Until this point, we are using the `Raise` DSL to combine different `Either`
+computations, with the goal of producing yet another `Either`. The frontier
+between both styles requires the use of `bind`; but if you go full-on with
+`Raise`, we can even remove those. In that case, the error type appears
+as the extension receiver, instead of wrapping the return type.
+
+```kotlin
+fun Raise<Error>.f(n: Int): String
+fun Raise<Error>.g(s: String): Thing
+fun Raise<Boo>.h(s: String): Thing
+fun Thing.summarize(): String
+```
+
+As a consequence of return types being "bare", we can use them directly,
+without the mediation of `bind`. The last of our examples reads now:
+
+```kotlin
+fun Raise<Error>.bar(n: Int): String {
+  val s = f(n)
+  val t = withError({ boo -> boo.toError() }) { h(s) }
+  return t.summarize()
+}
+```
+
+We encourage using this style, especially for non-public parts of your
+code, instead of continuously using `either` and `bind`. Apart from the
+stylistic improvement, it also avoids wrapping and unwrapping
+`Right` and `Left` values.
```

Comment:
OK, yikes.  I like the idea of not needing the bind() calls, but this is a case where I really want to see what's going on under the hood.  Because this example on its own just looks like magic, and I do not trust magic.

Reply:
I don't think that's important when using Raise; interested reader can consult the implementation.

In any case, Kotlin uses the same ideas of adding functionality via contexts in a lots of places. We use `launch` and `async` in `CoroutineScope` without having to understand the details of scheduling, task management, and so forth; we just trust the implementation and focus on the usage. This is my goal with this document.

Reply:
I am just offering the perspective of someone still new to Kotlin, but not to FP concepts. 😄  If the reader is not already very versed in other parts of Kotlin, this will be very confusing.